streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamSlidingWindowAggregateTest.java

```diff
@@ -100,17 +99,79 @@
 
     private final Properties props = StreamsTestUtils.getStreamsConfig(Serdes.String(), Serdes.String());
     private final String threadId = Thread.currentThread().getName();
+    private final String topic = "topic";
+    private final String defaultInOrderName = "InOrder";
+    private final String defaultReverseName = "Reverse";
+    private final long defaultWindowSize = 10L;
+    private final long defaultRetentionPeriod = 5000L;
+
+    private WindowBytesStoreSupplier getStoreSupplier(final boolean inOrderIterator,
+                                                      final String inOrderName,
+                                                      final String reverseName,
+                                                      final long windowSize) {
+        return inOrderIterator
+            ? new InOrderMemoryWindowStoreSupplier(inOrderName, defaultRetentionPeriod, windowSize, false)
+            : Stores.inMemoryWindowStore(reverseName, ofMillis(defaultRetentionPeriod), ofMillis(windowSize), false);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Test
+    public void testAggregateSmallInputWithZeroTimeDifference() {
+        final StreamsBuilder builder = new StreamsBuilder();
+
+        // We use CachingWindowStore to store the aggregated values internally, and then use TimeWindow to represent the "windowed KTable"
+        // thus, the window size must be greater than 0 here
+        final WindowBytesStoreSupplier storeSupplier = getStoreSupplier(inOrderIterator, defaultInOrderName, defaultReverseName, 1L);
```

Comment:
Cannot create a store supplier with window size of 0 here because we use `TimeWindow` to represent the "windowed KTable" result. Use window size of 1 instead. (Not sure if this should be an improvement or it is expected?)

Reply:
Are you saying the `CachingWindowStore` internally uses a `TimeWindow`? Or is the `TimeWindow` somewhere along the store supplier code path...?

Either way, doesn't this mean there's still a hole in the API since you can't use a custom WindowStore for a sliding windowed aggregation with the windowSize set to 0? If the WindowStore is going to represent different kinds of constant-size windows, it should probably be agnostic to the specific type of constant-sized window.

Reply:
> Are you saying the CachingWindowStore internally uses a TimeWindow? 

Yes, `CachingWindowStore` internally uses `TimeWindow` (i.e. `WindowKeySchema`). And looks like we use `TimeWinow` for `WindowStore`, and use `SessionWindow` for `SessionStore`.

> doesn't this mean there's still a hole in the API since you can't use a custom WindowStore for a sliding windowed aggregation with the windowSize set to 0?

I think so

> If the WindowStore is going to represent different kinds of constant-size windows, it should probably be agnostic to the specific type of constant-sized window.

Do you mean we should use `SessionWindow` (i.e. [start, end] inclusive time window) to represent the window? I'm not sure if this is the original design or just a miss. What do you think?

Reply:
> Do you mean we should use SessionWindow (i.e. [start, end] inclusive time window) to represent the window?

Not for the WindowStore (SessionWindow is fine for SessionStore of course), I'm saying if WindowStore is used for both inclusive-exclusive and also inclusive-inclusive windows, then we shouldn't ever assume one of them ie should not use an actual `TimeWindow` (or `InclusiveExclusiveWindow`) -- maybe we can just have a separate, plain window class that doesn't assume anything about its bounds.

We can even just (re)use the `TimeWindow` for this if we do the renaming as discussed, and move the check that prevents size = 0 to that new class. Then we can just have a plain data container class `TimeWindow` that does nothing but hold the start and end time for use in window-agnostic cases like the CachingWindowStore. Does that make sense?

Reply:
Make sense to me! I'll work on it. Thanks for suggestion!

Reply:
@mjsax , answering your review comments here. Please check the above discussion for more info.

> Why window-agnostic ? In general, I am not sure why we need to change the existing JavaDocs? What information do you think is missing or wrong?

> You add (inclusive) and (exclusive) in SessionWindow but remove it here. Seems inconsistent?

> Why do you remove this check? A TimeWindow should not allow this case.

> Why do we need to remove this temporarily?

--> The answer for the above questions are that we can't create a store supplier with window size of 0 here because we use `TimeWindow` to represent the "windowed KTable" result. @ableegoldman and I both thought it doesn't make sense to use `TimeWindow` to represent it if `WindowStore` is used for both inclusive-exclusive and also inclusive-inclusive windows. We should have a neutral time window for this case. That's why Sophie suggested that we should have a container class that does nothing but hold the start and end time for use in window-agnostic cases like the `CachingWindowStore`. And the container class can be named `TimeWindow`, and we were thinking that after all, we'll rename the `TimeWindow` into `InclusiveExclusiveWindow`, so that's why I changed the java doc/start and end time checking/test for it. 

So, since we agreed that we won't rename the window, I'll revert it. But still, there's a question there:  
_We can't create a store supplier with window size of 0 here because we use `TimeWindow` to represent the "windowed KTable" result._
Do you agree we should have a container class to do nothing but hold the start and end time for use in window-agnostic cases like the `CachingWindowStore`? Or any other suggestions?

Thank you.